cache/cache.go

```diff
@@ -61,76 +63,97 @@ func New[T any](expireAfter time.Duration) *Cache[T] {
 		panic("expireAfter duration must be positive")
 	}
 
-	return &Cache[T]{
-		data:        make(map[string]item[T]),
+	c := &Cache[T]{
+		data:        make(map[string]*cacheListItem[T]),
 		expireAfter: expireAfter,
+		stopCh:      make(chan struct{}),
 	}
-}
-
-// Removes an item by name and expiry time when the purge was scheduled.
-// If there is a race, and the item has been refreshed, it will not be purged.
-func (c *Cache[T]) purgeExpired(name string, expectedExpiryTime int64) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
 
-	if item, ok := c.data[name]; ok && item.expiresAt == expectedExpiryTime {
-		// found, and the expiry time is still the same as when the purge was requested.
-		delete(c.data, name)
+	// Start the sweep, with a minimum sweep of 50ms.
+	sweep := expireAfter / 4.0
+	if min := 50 * time.Millisecond; sweep < min {
+		sweep = min
 	}
+	go c.start(sweep)
+
+	return c
 }
 
-// Size returns the number of items in the cache.
+// Size returns the current number of items in the cache.
 func (c *Cache[T]) Size() int {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
+
+	if c.isStopped() {
+		panic("cache is stopped")
+	}
+
 	return len(c.data)
 }
 
-// Clear removes all items from the cache, regardless of their expiration.
+// Clear removes all items from the cache, regardless of their expiration. Note
+// this is different from Stop() which deletes all cached items and prevents new
+// items from being added.
 func (c *Cache[T]) Clear() {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.data = make(map[string]item[T])
+
+	if c.isStopped() {
```

Comment:
We have a lot of handling for stopping the cache. I'm a bit curious the use case we expect - i can't think of cases where i'd want to stop the cache without shutting down my application.

Reply:
Yea, you should really only terminate the cache if you're:

1. shutting down
2. creating an entirely new cache (e.g. to change the TTL)

Without this logic, you'd get really difficult to diagnose NPEs since we want to zero out as much of the old cache as possible to enable efficient GC. While still a panic, this is a consistent and debuggable panic.